ext/xsl/xsltprocessor.c

```diff
@@ -45,231 +45,83 @@ static zend_result php_xsl_xslt_apply_params(xsltTransformContextPtr ctxt, HashT
 	return SUCCESS;
 }
 
-static void xsl_ext_function_php(xmlXPathParserContextPtr ctxt, int nargs, int type) /* {{{ */
+static void xsl_proxy_factory(xmlNodePtr node, zval *child, dom_object *intern, xmlXPathParserContextPtr ctxt)
 {
-	xsltTransformContextPtr tctxt;
-	zval *args = NULL;
-	zval retval;
-	int i;
-	int error = 0;
-	zend_fcall_info fci;
-	zval handler;
-	xmlXPathObjectPtr obj;
-	char *str;
-	xsl_object *intern;
-	zend_string *callable = NULL;
-
+	ZEND_ASSERT(node->type != XML_NAMESPACE_DECL);
+
+	/**
+	 * Upon freeing libxslt's context, every document that is not the *main* document will be freed by libxslt.
+	 * If a node of a document that is *not the main* document gets returned to userland, we'd free the node twice:
+	 * first by the cleanup of the xslt context, and then by our own refcounting mechanism.
+	 * To prevent this, we'll take a copy if the node is not from the main document.
+	 * It is important that we do not copy the node unconditionally, because that means that:
+	 *  - modifications to the node will only modify the copy, and not the original
+	 *  - accesses to the parent, path, ... will not work
+	 */
```

Comment:
Well this is cursed and sad :(

Reply:
Yeah it is, it is pre-existing though not specific to this PR.